File: video-club-rental/python/tests/video_club_builder.py
from __future__ import annotations

from datetime import date

from video_club_rental import Title, User, VideoClub

from .fixed_clock import FixedClock
from .recording_notifier import RecordingNotifier


class VideoClubBuilder:
    def __init__(self) -> None:
        self._opened_on: date = date(2026, 1, 1)
        self._users: list[User] = []
        self._titles: list[Title] = []

    def opened_on(self, d: date) -> "VideoClubBuilder":
        self._opened_on = d
        return self

    def with_user(self, user: User) -> "VideoClubBuilder":
        self._users.append(user)
        return self

    def with_title(self, title: Title) -> "VideoClubBuilder":
        self._titles.append(title)
        return self

    def build(self) -> tuple[VideoClub, RecordingNotifier, FixedClock]:
        clock = FixedClock(self._opened_on)
        notifier = RecordingNotifier()
        club = VideoClub(clock, notifier)
        for user in self._users:
            club.seed_user(user)
        for title in self._titles:
            club.add_title(title)
        return club, notifier, clock

File: video-club-rental/typescript/tests/VideoClubBuilder.ts
import { Title } from '../src/Title.js';
import { User } from '../src/User.js';
import { VideoClub } from '../src/VideoClub.js';
import { FixedClock } from './FixedClock.js';
import { RecordingNotifier } from './RecordingNotifier.js';

export class VideoClubBuilder {
  private _openedOn: Date = new Date(Date.UTC(2026, 0, 1));
  private _users: User[] = [];
  private _titles: Title[] = [];

  openedOn(date: Date): this { this._openedOn = date; return this; }
  withUser(user: User): this { this._users.push(user); return this; }
  withTitle(title: Title): this { this._titles.push(title); return this; }

  build(): { club: VideoClub; notifier: RecordingNotifier; clock: FixedClock } {
    const clock = new FixedClock(this._openedOn);
    const notifier = new RecordingNotifier();
    const club = new VideoClub(clock, notifier);
    for (const u of this._users) club.seedUser(u);
    for (const t of this._titles) club.addTitle(t);
    return { club, notifier, clock };
  }
}
